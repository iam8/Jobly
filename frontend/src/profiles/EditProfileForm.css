.EditProfileForm {
    height: calc(100vh - 50px);
    display: flex;
    flex-direction: column;
    align-items: center;
}

.EditProfileForm-heading {
    padding-top: 3rem;
    padding-bottom: 1rem;
    color: white;
    text-shadow: 2px 2px 3px black;
}

.EditProfileForm-card {
    width: 30%;
}

.EditProfileForm label {
    font-weight: bold;
}