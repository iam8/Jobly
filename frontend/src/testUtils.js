import React from "react";
import UserContext from "./auth/UserContext";


const demoUser = {
    username: "testuser",
    firstName: "testfirst",
    lastName: "testlast",
    email: "test@test.net",
};


const DemoUserProvider = ({ children, currentUser = demoUser, hasAppliedToJob = () => false }) => (
    <UserContext.Provider value={{ currentUser, hasAppliedToJob }}>
        {children}
    </UserContext.Provider>
);


export { DemoUserProvider };
