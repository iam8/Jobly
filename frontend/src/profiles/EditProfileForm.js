import React, {useState, useContext} from "react";
import { Alert, Button, Card, CardBody, Form, FormGroup, Label, Input } from "reactstrap";

import JoblyApi from "../api";
import UserContext from "../auth/UserContext";

import "./EditProfileForm.css";


/**
 * Route: /profile
 *
 * Form used to edit user profile. Initially fills form inputs with the current user profile data.
 *
 * Manages state updates on changes to the form inputs.
 *
 * On form submission, attempts to update user profile with new data and displays a success or
 * failure notification.
 */
function EditProfileForm() {
    const {currentUser, setCurrentUser} = useContext(UserContext);
    const [saveSuccess, setSaveSuccess] = useState(false);
    const [formErrors, setFormErrors] = useState([]);

    const INIT_FORM = {
        firstName: currentUser.firstName,
        lastName: currentUser.lastName,
        email: currentUser.email
    };

    const [formData, setFormData] = useState(INIT_FORM);

    /** Handle changes to inputs. */
    const handleChange = (evt) => {
        const {name, value} = evt.target;
        setSaveSuccess(false);
        setFormErrors([]);
        setFormData((formData) => ({
            ...formData,
            [name]: value
        }));
    }

    /**
     * Call API to save new user profile data. Return saved data if successful or undefined if
     * unsuccessful.
     */
    async function saveUserProfile(username, formData) {
        try {
            const savedData = await JoblyApi.saveUserProfile({username, ...formData});
            return savedData;
        } catch(err) {
            setSaveSuccess(false);
            setFormErrors(err);
            console.log("ERROR SAVING PROFILE:", err);
            return;
        }
    }

    /** Form submission - attempt to update user profile data and reset form inputs. */
    const handleSubmit = async (evt) => {
        evt.preventDefault();
        const savedUserData = await saveUserProfile(currentUser.username, formData);
        if (!savedUserData) {
            return;
        }

        setCurrentUser(savedUserData);
        setSaveSuccess(true);
        setFormErrors([]);

        if ("password" in formData) {
            setFormData(data => ({
                ...data,
                password: ""
            }));
        }
    }

    /** Show an alert message on success or failure to update profile. */
    const renderAlert = () => {
        if (saveSuccess) {
            return (
                <Alert color="success">
                    Save successful!
                </Alert>
            )
        }

        if (formErrors.length) {
            return (
                <Alert color="danger">
                    Could not update profile - {formErrors}
                </Alert>
            );
        }
    }

    return (
        <div className="EditProfileForm">
            <h2 className="EditProfileForm-heading">Edit Profile</h2>

            <Card className="EditProfileForm-card">
                <CardBody>
                    <Form className="EditProfileForm-form" onSubmit={handleSubmit}>
                        <FormGroup>
                            <Label htmlFor="username">Username</Label>
                            <Input
                                id="username"
                                name="username"
                                type="text"
                                value={currentUser.username}
                                disabled
                            />
                        </FormGroup>

                        <FormGroup>
                            <Label htmlFor="first-name">First name</Label>
                            <Input
                                id="first-name"
                                name="firstName"
                                type="text"
                                value={formData.firstName}
                                onChange={handleChange}
                            />
                        </FormGroup>

                        <FormGroup>
                            <Label htmlFor="last-name">Last name</Label>
                            <Input
                                id="last-name"
                                name="lastName"
                                type="text"
                                value={formData.lastName}
                                onChange={handleChange}
                            />
                        </FormGroup>

                        <FormGroup>
                            <Label htmlFor="email">Email</Label>
                            <Input
                                id="email"
                                name="email"
                                type="email"
                                value={formData.email}
                                onChange={handleChange}
                            />
                        </FormGroup>

                        {renderAlert()}

                        <Button color="primary" block>Save changes</Button>
                    </Form>
                </CardBody>
            </Card>
        </div>
    )
}


export default EditProfileForm;
