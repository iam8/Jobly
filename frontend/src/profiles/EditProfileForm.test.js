import React from "react";
import { render } from "@testing-library/react";
import { DemoUserProvider as UserProvider } from "../testUtils";
import EditProfileForm from "./EditProfileForm";


test("Renders without crashing", () => {
    render (
        <UserProvider>
            <EditProfileForm />
        </UserProvider>
    );
})


test("Matches snapshot", () => {
    const { asFragment } = render(
        <UserProvider>
            <EditProfileForm />
        </UserProvider>
    );

    expect(asFragment()).toMatchSnapshot();
});