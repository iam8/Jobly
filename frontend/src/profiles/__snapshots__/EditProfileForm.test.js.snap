// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="EditProfileForm"
  >
    <h2
      class="EditProfileForm-heading"
    >
      Edit Profile
    </h2>
    <div
      class="EditProfileForm-card card"
    >
      <div
        class="card-body"
      >
        <form
          class="EditProfileForm-form"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="username"
            >
              Username
            </label>
            <input
              class="form-control"
              disabled=""
              id="username"
              name="username"
              type="text"
              value="testuser"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="first-name"
            >
              First name
            </label>
            <input
              class="form-control"
              id="first-name"
              name="firstName"
              type="text"
              value="testfirst"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="last-name"
            >
              Last name
            </label>
            <input
              class="form-control"
              id="last-name"
              name="lastName"
              type="text"
              value="testlast"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="email"
            >
              Email
            </label>
            <input
              class="form-control"
              id="email"
              name="email"
              type="email"
              value="test@test.net"
            />
          </div>
          <button
            class="btn btn-primary d-block w-100"
          >
            Save changes
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
